CONTRIBUTING.md

```diff
@@ -1,57 +1,169 @@
+still recommended to follow the rules above to make the life of maintainers
+easier.
+
+If you want to communicate something to reviewers, but it doesn't belong to
+the commit message, write that under `--` on PR body. Your PR could look like:
+
+  ```
+  TITLE (matches commit title)
+
+  Body (matches commit body)
+
+  --
+
+  Your arbitrary message goes here.
+
+  Various checkboxes PR template required you to fill.
+  ```
+
+#### Handling review
+
+During the review process certain events could require an action from your side,
+to communication more efficient the common patterns and reactions are described
+below.
+
+_Event:_ The CI fails to build, but it looks like not your fault. Not
+communicating so, could result into maintainers not looking into your patch,
+unless they CI that CI pass.\
+_Desired behavior:_ Write a message saying roughly the following "The CI failure
+is unrelated", so maintainers will fix it for you.
+
+_Event:_ Collaborator requested review on your PR.\
+_Desired behavior:_ Once you address the request, you _should_ re-request review
+with the github's UI. If you don't agree with what maintainer suggested, you
+should object that and re-request the review. That will indicate that the
+_ball_ is on maintainer's side.
+
+_Event:_ You've opened a PR, but maintainer shortly after commented that they
+want to work on that _themselves_.\
+_Desired behavior:_ Discusses with maintainer their plans if they were not
+outlined in the initial response, because such response means that they
+are not interested in reviewing your code. Such thing could happen when
+underestimating complexity of the task you're solving or when your patch
+mandate certain downstream designs.
+
+[prs]: https://github.com/rust-windowing/winit/pulls
+[issues]: https://github.com/rust-windowing/winit/issues
+[matrix]: https://matrix.to/#/#rust-windowing:matrix.org
 
 ## Maintainers
 
-The current maintainers for each platform are listed in the [CODEOWNERS](.github/CODEOWNERS) file.
-
-## Release process
+Winit has plenty of maintainers with different backgrounds, experience level,
+and reasons to be winit maintainer in the first place. To ensure that winit's
+code quality is not decreasing over time and to make it easier to teach new
+maintainers the _winit way of doing things_ the common policies and routines
+are defined in this section.
+
+The current maintainers for each platform are listed in [this file][CODEOWNERS].
+
+### Contributions handling
+
+The maintainer must ensure that the external contributions meet the winit's
+quality standards. If it's not, it **is maintainer's responsibility** to bring
+it on par, which includes:
+
+  - Decline the patch if it tries to achieve short term goals instead of solving
+    the problem in general. In such case maintainer must communicate that **as
+    early as possible**, so contributors don't spend their time on something
+    that won't be accepted in the end.
```

Comment:
I acknowledge that this may lead to decreased code quality, but I believe that requiring the problem to be solved in general is the wrong approach for Winit. While the project is a few years old, it's still young enough that we don't yet know how to do everything! And the absolute best way to get knowledge is to let people experiment with the code!

As an example, https://github.com/rust-windowing/winit/pull/3499 is solving a short-term goal, but I believe it'll also help guide how the API will have to look like in the future.

Another example is the file open APIs on macOS, which got rejected partly because of technical difficulties in making it work right, but also because we wanted to find some solution to https://github.com/rust-windowing/winit/issues/2120, which has been open for over two years with only recently a solution on the horizon.

My opinion is that we need to get all of the sub-par code into Winit, so that we can iterate on it in there instead of in a PR that will have to be rebased over an over (keyboard v2).

Reply:
>I acknowledge that this may lead to decreased code quality, but I believe that requiring the problem to be solved in general is the wrong approach for Winit. While the project is a few years old, it's still young enough that we don't yet know how to do everything! And the absolute best way to get knowledge is to let people experiment with the code!

it's like 10 years old at this point and it was inside glutin at some point.

>As an example, https://github.com/rust-windowing/winit/pull/3499 is solving a short-term goal, but I believe it'll also help guide how the API will have to look like in the future.

This is not how the API will look though. The trait extension is _fine_, but it was not proposed there.

>Another example is the file open APIs on macOS, which got rejected partly because of technical difficulties in making it work right, but also because we wanted to find some solution to https://github.com/rust-windowing/winit/issues/2120, which has been open for over two years with only recently a solution on the horizon.

Yes, the solution is merged and can be used to implement the said API already.

>My opinion is that we need to get all of the sub-par code into Winit, so that we can iterate on it in there instead of in a PR that will have to be rebased over an over (keyboard v2).

No, the subpar code will stay with us forever, that's how it works unfortunately. Once you land it it'll be pretty much like that for a long time. There're exceptions, but it's how it is, unfortunately. You should never justify not maintainable code.